### PR TITLE
Feature/point source markers on imi preview

### PR DESCRIPTION
**Pull request by alexoort**
### Name and Institution (Required)

Name: Alex Oort Alonso
Institution: Harvard

### Describe the update

Adds markers for any specified native resolution coordinates on preview graphs. Integrates weekly SRON plumes into the clustering algorithm. Some other small changes (i.e. add notebook that allows users to map their statevector files, and notebook that allows users to merge several shapefiles together.)

### CODE REVIEW

File: src/notebooks/visualization_notebook.ipynb [hunk 1]
```diff
@@ -187,6 +187,52 @@
     "print('Posterior emissions :', total_posterior_emissions, 'Tg/y')"
    ]
   },
+  {
```
Comment (laestrada) [inline src/notebooks/visualization_notebook.ipynb:190]: can we make a separate PR for the sectoral emission attribution to avoid confusion?

**Review by laestrada — CHANGES_REQUESTED** [review]
Nice work Alex! Impressed by your web scraper code. Additionally, these changes should probably be made in a fork of the `dev-2.0` branch so that the PR can be directly applied to that. As is, this relies on the feature/kalman-filter-refactor branch being merged into dev-2.0 first. Don't worry about the branches for now though. I think the kalman-filter-refactor branch will be merged soon.

Some general things to do for all your code:
- add additional comments. we expect scientists to poke around the code often, so having good documentation is crucial
- add header comments to all new functions describing the purpose of the function, arguments, and return type

I also added many comments with specifics on how to improve your pull request before it gets approved.

File: src/components/statevector_component/aggregation.py [hunk 2]
```diff
@@ -295,6 +301,93 @@ def generate_cluster_pairs(config, sensitivities):
...
+    polygons = shpfilePoints
+    for lon, lat in zip(plumes['lon'], plumes['lat']):   
+        inShape = False
+        point = Point(lon, lat)
+        for polygon in polygons:
+            poly = Polygon(polygon)
+            if poly.contains(point):
+                inShape = True
+        if not inShape:
+            plumes  = plumes[(plumes['lon'] != float(lon)) | (plumes['lat'] != float(lat))]
```
[the remark this answers is not in the document]
Comment (alexoort) [reply src/components/statevector_component/aggregation.py:335]: Yes!

**Comment by alexoort** [comment]
Made suggested changes!

File: src/components/statevector_component/aggregation.py [hunk 3]
```diff
@@ -343,7 +460,15 @@ def force_native_res_pixels(config, clusters, sensitivities):
         cluster_pairs    [(tuple)]: cluster pairings
     Returns:             [double] : updated sensitivities
     """
+    plumes = SRON_plumes(config)
```
Comment (laestrada) [inline src/components/statevector_component/aggregation.py:463]: This should be an option rather than happening automatically. Maybe we need another config variable eg. `PointSourceDatasets: ["SRON"]`. Making it a list allows for easier extension to other public datasources in the future eg. `IMEO`

File: src/components/statevector_component/aggregation.py [hunk 4]
```diff
@@ -296,6 +304,115 @@ def generate_cluster_pairs(config, sensitivities):
...
+            csvUrl = url + "/" + link.get('href')
+            dates = csvUrl.split("_v")[1]
+            try:
+                rcsv = requests.get(csvUrl, allow_redirects=True)
+                file = f"SRON_{dates}"
+                open(file, 'wb').write(rcsv.content) # downloads all of the plumes from that week in a CSV file in the current directory
+                df = pd.read_csv(file) # reads from the csv file into a pandas dataframe
+                plume = plume.append(df, ignore_index=True)
+            except:
+                print("ERROR FETCHING CSV FILE")
```
Comment (laestrada) [inline src/components/statevector_component/aggregation.py:332]: Update error message to say: `f"Warning: Unable to access data for csv file at {csvUrl}. The file may not exist or there may be a connection problem."`

**Review by laestrada — COMMENTED** [review]
This looks great Alex! Two nitpick comments and then I think it will be all set. Did you test it with the updates from my last review?

File: src/components/statevector_component/aggregation.py [hunk 5]
```diff
@@ -343,7 +460,15 @@ def force_native_res_pixels(config, clusters, sensitivities):
...
     Returns:             [double] : updated sensitivities
     """
+    plumes = SRON_plumes(config)
     coords = read_coordinates(config["ForcedNativeResolutionElements"])
+    if plumes is not None:
+        if coords is None:
+            coords = plumes
+        else:
+            coords.extend(plumes)
+        if (len(coords)) > config["NumberOfElements"]:
```
Comment (laestrada) [inline src/components/statevector_component/aggregation.py:470]: I just realized his should only be applied to coordinates with lats and lons that are in different grid cells, since there can be multiple point sources in the same grid cell. I think all you need to do is use the binned lon and binned lats instead (defined lower down), making sure to drop any coords that are the same once binned.

File: src/components/statevector_component/aggregation.py [hunk 6]
```diff
@@ -399,15 +398,18 @@ def SRON_plumes(config):
...
+            p = get_plumes(str(currentDate.month), str(currentDate.year))
+            plumes = pd.concat([plumes, pd.DataFrame(p)], ignore_index=True) 
         currentDate = currentDate + relativedelta(months=1)
     
     #filters through the dataset to remove any plumes outside the ROI
     if custom_vectorfile:
         plumes = shapefile_filter(plumes, shapefile_path) #calls function to filter through coordinates found in shapefile
     else:
         plumes = rectangular_filter(plumes, LatMax, LatMin, LonMax, LonMin)
+    print("Detected plumes: ")
```
Comment (laestrada) [inline src/components/statevector_component/aggregation.py:411]: This looks like it was printed for testing purposes? Maybe remove

File: src/components/statevector_component/aggregation.py [hunk 7]
```diff
@@ -480,11 +489,26 @@ def force_native_res_pixels(config, clusters, sensitivities):
...
         lat_step = 0.5
         lon_step = 0.625
+    
+    for lat, lon in coords:
+        lon = np.floor(lon / lon_step) * lon_step
+        lat = np.floor(lat / lat_step) * lat_step
+
+ 
+    print(f"Length of the list before: {len(coords)}")
+    coords = sorted(set(map(tuple, coords)), reverse=True) # Remove any duplicate coordinates within the same gridcell.
```
Comment (laestrada) [inline src/components/statevector_component/aggregation.py:499]: You may consider using [numpy.unique](https://numpy.org/doc/stable/reference/generated/numpy.unique.html) for this.

What does map do here?

File: src/components/statevector_component/aggregation.py [hunk 8]
```diff
@@ -480,11 +489,26 @@ def force_native_res_pixels(config, clusters, sensitivities):
...
+    for lat, lon in coords:
+        lon = np.floor(lon / lon_step) * lon_step
+        lat = np.floor(lat / lat_step) * lat_step
+
+ 
+    print(f"Length of the list before: {len(coords)}")
+    coords = sorted(set(map(tuple, coords)), reverse=True) # Remove any duplicate coordinates within the same gridcell. 
+    coords = [list(coordinate) for coordinate in coords]
+    print(coords)
+    print(f"Length of the list after: {len(coords)}")
```
Comment (laestrada) [inline src/components/statevector_component/aggregation.py:502]: remove the print statements ;)

File: src/components/statevector_component/aggregation.py [hunk 9]
```diff
@@ -480,11 +489,26 @@ def force_native_res_pixels(config, clusters, sensitivities):
...
+
+    if (len(coords)) > config["NumberOfElements"]:
+                coords = coords[0:config["NumberOfElements"] - 1]
```
Comment (laestrada) [inline src/components/statevector_component/aggregation.py:505]: the indentation here looks messed up

File: src/components/statevector_component/aggregation.py [hunk 10]
```diff
@@ -480,11 +489,26 @@ def force_native_res_pixels(config, clusters, sensitivities):
...
+        lat = np.floor(lat / lat_step) * lat_step
+
+ 
+    print(f"Length of the list before: {len(coords)}")
+    coords = sorted(set(map(tuple, coords)), reverse=True) # Remove any duplicate coordinates within the same gridcell. 
+    coords = [list(coordinate) for coordinate in coords]
+    print(coords)
+    print(f"Length of the list after: {len(coords)}")
+
+    if (len(coords)) > config["NumberOfElements"]:
```
Comment (laestrada) [inline src/components/statevector_component/aggregation.py:504]: I dont think you need the parentheses around len(coords)

File: src/components/statevector_component/aggregation.py [hunk 11]
```diff
@@ -460,18 +462,25 @@ def force_native_res_pixels(config, clusters, sensitivities):
...
         else:
             coords.extend(plumes)
-        if (len(coords)) > config["NumberOfElements"]:
-                coords = coords[0:config["NumberOfElements"] - 1]
+
+        
 
     if coords is None:
         # No forced pixels inputted
+        print("No forced native pixels")
```
Comment (laestrada) [inline src/components/statevector_component/aggregation.py:483]: Maybe clarify "No forced native pixels in {Source} dataset."

**Review by laestrada — COMMENTED** [review]
A few more minor comments

File: src/components/statevector_component/aggregation.py (hunk 7 above)
Comment (alexoort) [reply src/components/statevector_component/aggregation.py:499]: Since the set() function cannot be applied to mutable objects (e.g. lists), I converted the coordinates into tuples using the map() function.

**Review by laestrada — APPROVED** [review]
Nice work Alex, PR approved! I will merge it into the `dev-2.0` branch after the `kalman-filter-refactor` branch gets merged in
